processor/tailsamplingprocessor/sampling/string_tag_filter.go

```diff
@@ -15,21 +15,67 @@
 package sampling
 
 import (
+	"regexp"
+
+	"github.com/golang/groupcache/lru"
 	"go.opentelemetry.io/collector/consumer/pdata"
 	"go.uber.org/zap"
 )
 
+const defaultCacheSize = 128
+
 type stringAttributeFilter struct {
 	key    string
-	values map[string]struct{}
 	logger *zap.Logger
+	// matcher defines the func to match the attribute values in strict string
+	// or in regular expression
+	matcher func(string) bool
+}
+
+type regexStrSetting struct {
+	matchedAttrs *lru.Cache
+	filterList   []*regexp.Regexp
 }
 
 var _ PolicyEvaluator = (*stringAttributeFilter)(nil)
 
 // NewStringAttributeFilter creates a policy evaluator that samples all traces with
 // the given attribute in the given numeric range.
-func NewStringAttributeFilter(logger *zap.Logger, key string, values []string) PolicyEvaluator {
+func NewStringAttributeFilter(logger *zap.Logger, key string, values []string, regexMatchEnabled bool, evictSize int) PolicyEvaluator {
+	// initialize regex filter rules and LRU cache for matched results
+	if regexMatchEnabled {
+		if evictSize <= 0 {
+			evictSize = defaultCacheSize
+		}
+		filterList := addFilters(values)
+		regexStrSetting := &regexStrSetting{
+			matchedAttrs: lru.New(evictSize),
+			filterList:   filterList,
+		}
+
+		return &stringAttributeFilter{
+			key:    key,
+			logger: logger,
+			// matcher returns true if the given string matches the regex rules defined in string attribute filters
+			matcher: func(toMatch string) bool {
```

Comment:
If you keep this logic isolated in its own function, you can test it more easily.

Reply:
The anonymous funcs take the captured variable initialized in the constructor. I want to leave it as is. Thanks.